File: bindings/python/python/openpit/account_adjustment.py
# Copyright The Pit Project Owners. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Please see https://github.com/openpitkit and the OWNERS file for details.

from __future__ import annotations

import typing

from ._openpit import AccountAdjustment as _AccountAdjustment
from ._openpit import AccountAdjustmentAmount as _AccountAdjustmentAmount
from ._openpit import (
    AccountAdjustmentBalanceOperation as _AccountAdjustmentBalanceOperation,
)
from ._openpit import AccountAdjustmentBounds as _AccountAdjustmentBounds
from ._openpit import (
    AccountAdjustmentPositionOperation as _AccountAdjustmentPositionOperation,
)
from .core import Instrument
from .param import (
    AdjustmentAmount,
    Asset,
    Leverage,
    PositionMode,
    PositionSize,
    Price,
)


def _require_instance(
    value: typing.Any,
    expected_type: type[typing.Any],
    *,
    name: str,
) -> typing.Any:
    if value is None:
        return None
    if not isinstance(value, expected_type):
        raise TypeError(
            f"{name} must be {expected_type.__module__}.{expected_type.__name__}"
        )
    return value


class Amount(_AccountAdjustmentAmount):
    """Grouped balance/held/incoming adjustment payload."""

    # @typing.override
    def __new__(cls, *args: typing.Any, **kwargs: typing.Any) -> Amount:
        _ = args, kwargs
        return _AccountAdjustmentAmount.__new__(cls)

    # @typing.override
    def __init__(
        self,
        *,
        balance: AdjustmentAmount | None = None,
        held: AdjustmentAmount | None = None,
        incoming: AdjustmentAmount | None = None,
    ) -> None:
        _AccountAdjustmentAmount.balance.__set__(self, balance)
        _AccountAdjustmentAmount.held.__set__(self, held)
        _AccountAdjustmentAmount.incoming.__set__(self, incoming)

    # @typing.override
    @property
    def balance(self) -> AdjustmentAmount | None:
        """Actual resulting balance/position value."""
        value = _AccountAdjustmentAmount.balance.__get__(self, type(self))
        if value is None:
            return None
        if value.is_delta:
            return AdjustmentAmount.delta(value.as_delta)
        return AdjustmentAmount.absolute(value.as_absolute)

    # @typing.override
    @property
    def held(self) -> AdjustmentAmount | None:
        """Amount earmarked for outgoing settlement.

        Unavailable for immediate use.
        """
        value = _AccountAdjustmentAmount.held.__get__(self, type(self))
        if value is None:
            return None
        if value.is_delta:
            return AdjustmentAmount.delta(value.as_delta)
        return AdjustmentAmount.absolute(value.as_absolute)

    # @typing.override
    @property
    def incoming(self) -> AdjustmentAmount | None:
        """Amount in-flight for incoming acquisition and not yet finalized."""
        value = _AccountAdjustmentAmount.incoming.__get__(self, type(self))
        if value is None:
            return None
        if value.is_delta:
            return AdjustmentAmount.delta(value.as_delta)
        return AdjustmentAmount.absolute(value.as_absolute)

    def __repr__(self) -> str:
        return _AccountAdjustmentAmount.__repr__(self)


class BalanceOperation(_AccountAdjustmentBalanceOperation):
    """Direct physical balance adjustment."""

    # @typing.override
    def __new__(cls, *args: typing.Any, **kwargs: typing.Any) -> BalanceOperation:
        _ = args, kwargs
        return _AccountAdjustmentBalanceOperation.__new__(cls)

    # @typing.override
    def __init__(
        self,
        *,
        asset: Asset,
        average_entry_price: Price | None = None,
    ) -> None:
        _AccountAdjustmentBalanceOperation.asset.__set__(self, asset)
        _AccountAdjustmentBalanceOperation.average_entry_price.__set__(
            self, average_entry_price
        )

    # @typing.override
    @property
    def asset(self) -> Asset:
        """Adjusted balance asset."""
        return _AccountAdjustmentBalanceOperation.asset.__get__(self, type(self))

    # @typing.override
    @property
    def average_entry_price(self) -> Price | None:
        """Optional cost basis for the adjusted physical balance."""
        value = _AccountAdjustmentBalanceOperation.average_entry_price.__get__(
            self, type(self)
        )
        return value

    def __repr__(self) -> str:
        return _AccountAdjustmentBalanceOperation.__repr__(self)


class PositionOperation(_AccountAdjustmentPositionOperation):
    """Direct derivatives-like position adjustment."""

    # @typing.override
    def __new__(cls, *args: typing.Any, **kwargs: typing.Any) -> PositionOperation:
        _ = args, kwargs
        return _AccountAdjustmentPositionOperation.__new__(cls)

    # @typing.override
    def __init__(
        self,
        *,
        instrument: Instrument,
        collateral_asset: Asset,
        average_entry_price: Price,
        mode: PositionMode,
        leverage: Leverage | int | float | None = None,
    ) -> None:
        # Structural aggregate check is intentionally kept in Python so
        # aggregate misuse fails with a clear contract error instead of
        # indirect attribute/field errors during aggregation.
        _require_instance(instrument, Instrument, name="instrument")
        _AccountAdjustmentPositionOperation.underlying_asset.__set__(
            self, instrument.underlying_asset
        )
        _AccountAdjustmentPositionOperation.settlement_asset.__set__(
            self, instrument.settlement_asset
        )
        _AccountAdjustmentPositionOperation.collateral_asset.__set__(
            self, collateral_asset
        )
        _AccountAdjustmentPositionOperation.average_entry_price.__set__(
            self, average_entry_price
        )
        _AccountAdjustmentPositionOperation.mode.__set__(self, mode)
        _AccountAdjustmentPositionOperation.leverage.__set__(self, leverage)
        self.__dict__["_py_instrument"] = instrument

    @property
    def instrument(self) -> Instrument:
        """Adjusted position instrument."""
        return self.__dict__["_py_instrument"]

    # @typing.override
    @property
    def collateral_asset(self) -> Asset:
        """Collateral asset used by the adjusted position."""
        return _AccountAdjustmentPositionOperation.collateral_asset.__get__(
            self, type(self)
        )

    # @typing.override
    @property
    def average_entry_price(self) -> Price:
        """Average entry price for the adjusted position state."""
        return _AccountAdjustmentPositionOperation.average_entry_price.__get__(
            self, type(self)
        )

    # @typing.override
    @property
    def mode(self) -> PositionMode:
        """Netting vs hedged position representation."""
        return _AccountAdjustmentPositionOperation.mode.__get__(self, type(self))

    # @typing.override
    @property
    def leverage(self) -> Leverage | None:
        """Optional leverage snapshot/setting carried with the position adjustment."""
        return _AccountAdjustmentPositionOperation.leverage.__get__(self, type(self))

    def __repr__(self) -> str:
        return _AccountAdjustmentPositionOperation.__repr__(self)


class Bounds(_AccountAdjustmentBounds):
    """Optional post-adjustment inclusive limits."""

    # @typing.override
    def __new__(cls, *args: typing.Any, **kwargs: typing.Any) -> Bounds:
        _ = args, kwargs
        return _AccountAdjustmentBounds.__new__(cls)

    # @typing.override
    def __init__(
        self,
        *,
        balance_upper: PositionSize | None = None,
        balance_lower: PositionSize | None = None,
        held_upper: PositionSize | None = None,
        held_lower: PositionSize | None = None,
        incoming_upper: PositionSize | None = None,
        incoming_lower: PositionSize | None = None,
    ) -> None:
        _AccountAdjustmentBounds.balance_upper.__set__(self, balance_upper)
        _AccountAdjustmentBounds.balance_lower.__set__(self, balance_lower)
        _AccountAdjustmentBounds.held_upper.__set__(self, held_upper)
        _AccountAdjustmentBounds.held_lower.__set__(self, held_lower)
        _AccountAdjustmentBounds.incoming_upper.__set__(self, incoming_upper)
        _AccountAdjustmentBounds.incoming_lower.__set__(self, incoming_lower)

    @property
    def balance_upper(self) -> PositionSize | None:
        """Allowed post-adjustment inclusive upper bound for balance."""
        return _AccountAdjustmentBounds.balance_upper.__get__(self, type(self))

    @property
    def balance_lower(self) -> PositionSize | None:
        """Allowed post-adjustment inclusive lower bound for balance."""
        return _AccountAdjustmentBounds.balance_lower.__get__(self, type(self))

    @property
    def held_upper(self) -> PositionSize | None:
        """Allowed post-adjustment inclusive upper bound for held."""
        return _AccountAdjustmentBounds.held_upper.__get__(self, type(self))

    @property
    def held_lower(self) -> PositionSize | None:
        """Allowed post-adjustment inclusive lower bound for held."""
        return _AccountAdjustmentBounds.held_lower.__get__(self, type(self))

    @property
    def incoming_upper(self) -> PositionSize | None:
        """Allowed post-adjustment inclusive upper bound for incoming."""
        return _AccountAdjustmentBounds.incoming_upper.__get__(self, type(self))

    @property
    def incoming_lower(self) -> PositionSize | None:
        """Allowed post-adjustment inclusive lower bound for incoming."""
        return _AccountAdjustmentBounds.incoming_lower.__get__(self, type(self))

    def __repr__(self) -> str:
        return _AccountAdjustmentBounds.__repr__(self)


class Adjustment(_AccountAdjustment):
    """Extensible non-trading account-adjustment model.

    Snapshot semantics:
    On ``Engine.apply_account_adjustment(...)``, each adjustment item is
    snapshotted at submission time for policy evaluation. Mutations of
    adjustment objects after submission do not affect the in-flight batch.
    """

    # @typing.override
    def __new__(cls, *args: typing.Any, **kwargs: typing.Any) -> Adjustment:
        _ = args, kwargs
        return _AccountAdjustment.__new__(cls)

    # @typing.override
    def __init__(
        self,
        *,
        operation: BalanceOperation | PositionOperation | None = None,
        amount: Amount | None = None,
        bounds: Bounds | None = None,
    ) -> None:
        # Structural aggregate check is intentionally kept in Python because
        # operation is a Python-only union wrapper at the public boundary.
        if operation is not None and not isinstance(
            operation,
            (BalanceOperation, PositionOperation),
        ):
            raise TypeError(
                "operation must be "
                "openpit.account_adjustment.BalanceOperation or "
                "openpit.account_adjustment.PositionOperation"
            )
        # Structural checks for aggregate groups stay at Python boundary to keep
        # explicit API-contract errors for wrong wrapper types.
        _require_instance(amount, Amount, name="amount")
        _require_instance(bounds, Bounds, name="bounds")
        _AccountAdjustment.operation.__set__(self, operation)
        _AccountAdjustment.amount.__set__(self, amount)
        _AccountAdjustment.bounds.__set__(self, bounds)
        self.__dict__["_py_operation"] = operation
        self.__dict__["_py_amount"] = amount
        self.__dict__["_py_bounds"] = bounds

    @property
    def operation(
        self,
    ) -> BalanceOperation | PositionOperation | None:
        """Adjustment operation details group."""
        return self.__dict__.get("_py_operation")

    @property
    def amount(self) -> Amount | None:
        """Adjustment amount deltas group."""
        return self.__dict__.get("_py_amount")

    @property
    def bounds(self) -> Bounds | None:
        """Optional post-adjustment bounds group."""
        return self.__dict__.get("_py_bounds")

    def __repr__(self) -> str:
        return _AccountAdjustment.__repr__(self)


__all__ = [
    "Adjustment",
    "Amount",
    "BalanceOperation",
    "Bounds",
    "PositionOperation",
]

File: crates/openpit/src/core/account_adjustment.rs
// Copyright The Pit Project Owners. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Please see https://github.com/openpitkit and the OWNERS file for details.

use crate::param::{AdjustmentAmount, Asset, Leverage, PositionMode, PositionSize, Price};
use crate::{impl_request_has_field, impl_request_has_field_passthrough};

use super::{
    HasAccountAdjustmentBalance, HasAccountAdjustmentBalanceAverageEntryPrice,
    HasAccountAdjustmentBalanceLowerBound, HasAccountAdjustmentBalanceUpperBound,
    HasAccountAdjustmentHeld, HasAccountAdjustmentHeldLowerBound,
    HasAccountAdjustmentHeldUpperBound, HasAccountAdjustmentIncoming,
    HasAccountAdjustmentIncomingLowerBound, HasAccountAdjustmentIncomingUpperBound,
    HasAccountAdjustmentPositionLeverage, HasAverageEntryPrice, HasBalanceAsset,
    HasCollateralAsset, HasPositionInstrument, HasPositionMode, Instrument,
};

/// Grouped balance/held/incoming adjustment payload.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct AccountAdjustmentAmount {
    /// Settled balance/position after applying the adjustment.
    /// Includes only fully finalized amounts. Cash balance for cash accounts,
    /// position for instruments.
    pub balance: Option<AdjustmentAmount>,

    /// Portion of `balance` reserved against future outgoing flow and
    /// unavailable for new commitments.
    /// Covers both:
    /// - working orders that, if filled, would deduct from balance;
    /// - filled trades awaiting outgoing settlement (T+N).
    pub held: Option<AdjustmentAmount>,

    /// Expected future inflow into `balance`, not yet finalized.
    /// Covers both:
    /// - working orders that, if filled, would add to balance;
    /// - filled trades awaiting incoming settlement (T+N).
    pub incoming: Option<AdjustmentAmount>,
}

/// Adds grouped balance/held/incoming adjustment payload.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct WithAccountAdjustmentAmount<T> {
    pub inner: T,
    pub amount: AccountAdjustmentAmount,
}

impl_request_has_field!(
    AccountAdjustmentAmount,
    WithAccountAdjustmentAmount,
    amount,
    HasAccountAdjustmentBalance, balance, Option<AdjustmentAmount>, balance;
    HasAccountAdjustmentHeld, held, Option<AdjustmentAmount>, held;
    HasAccountAdjustmentIncoming, incoming, Option<AdjustmentAmount>, incoming;
);
impl_request_has_field_passthrough!(
    WithAccountAdjustmentAmount,
    inner,
    HasBalanceAsset, balance_asset, &Asset;
    HasAccountAdjustmentBalanceAverageEntryPrice, balance_average_entry_price, Option<Price>;
    HasPositionInstrument, position_instrument, &Instrument;
    HasCollateralAsset, collateral_asset, &Asset;
    HasAverageEntryPrice, average_entry_price, Price;
    HasPositionMode, position_mode, PositionMode;
    HasAccountAdjustmentPositionLeverage, position_leverage, Option<Leverage>;
    HasAccountAdjustmentBalanceUpperBound, balance_upper, Option<PositionSize>;
    HasAccountAdjustmentBalanceLowerBound, balance_lower, Option<PositionSize>;
    HasAccountAdjustmentHeldUpperBound, held_upper, Option<PositionSize>;
    HasAccountAdjustmentHeldLowerBound, held_lower, Option<PositionSize>;
    HasAccountAdjustmentIncomingUpperBound, incoming_upper, Option<PositionSize>;
    HasAccountAdjustmentIncomingLowerBound, incoming_lower, Option<PositionSize>;
);

/// Direct adjustment of a physical asset balance without hedge/netting semantics.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct AccountAdjustmentBalanceOperation {
    pub asset: Asset,
    /// Optional cost basis for the adjusted physical balance.
    pub average_entry_price: Option<Price>,
}

/// Adds physical-balance adjustment operation payload.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct WithAccountAdjustmentBalanceOperation<T> {
    pub inner: T,
    pub operation: AccountAdjustmentBalanceOperation,
}

impl_request_has_field!(
    AccountAdjustmentBalanceOperation,
    WithAccountAdjustmentBalanceOperation,
    operation,
    HasBalanceAsset, balance_asset, &Asset, asset;
);
impl_request_has_field!(
    AccountAdjustmentBalanceOperation,
    WithAccountAdjustmentBalanceOperation,
    operation,
    HasAccountAdjustmentBalanceAverageEntryPrice, balance_average_entry_price, Option<Price>, average_entry_price;
);
impl_request_has_field_passthrough!(
    WithAccountAdjustmentBalanceOperation,
    inner,
    HasAverageEntryPrice, average_entry_price, Price;
    HasAccountAdjustmentBalance, balance, Option<AdjustmentAmount>;
    HasAccountAdjustmentHeld, held, Option<AdjustmentAmount>;
    HasAccountAdjustmentIncoming, incoming, Option<AdjustmentAmount>;
    HasAccountAdjustmentBalanceUpperBound, balance_upper, Option<PositionSize>;
    HasAccountAdjustmentBalanceLowerBound, balance_lower, Option<PositionSize>;
    HasAccountAdjustmentHeldUpperBound, held_upper, Option<PositionSize>;
    HasAccountAdjustmentHeldLowerBound, held_lower, Option<PositionSize>;
    HasAccountAdjustmentIncomingUpperBound, incoming_upper, Option<PositionSize>;
    HasAccountAdjustmentIncomingLowerBound, incoming_lower, Option<PositionSize>;
    HasPositionInstrument, position_instrument, &Instrument;
    HasCollateralAsset, collateral_asset, &Asset;
    HasPositionMode, position_mode, PositionMode;
    HasAccountAdjustmentPositionLeverage, position_leverage, Option<Leverage>;
);

/// Direct adjustment of a derivatives-like position with explicit position mode.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct AccountAdjustmentPositionOperation {
    pub instrument: Instrument,
    /// Asset used to collateralize and settle the adjusted position state.
    ///
    /// This is the margin/collateral bucket affected by the adjustment, not
    /// the traded underlying asset itself.
    pub collateral_asset: Asset,
    /// Average entry price for the adjusted position state.
    pub average_entry_price: Price,
    /// Netting vs hedged position representation.
    pub mode: PositionMode,
    /// Optional leverage snapshot/setting carried with the position adjustment.
    pub leverage: Option<Leverage>,
}

/// Adds derivatives-position adjustment operation payload.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct WithAccountAdjustmentPositionOperation<T> {
    pub inner: T,
    pub operation: AccountAdjustmentPositionOperation,
}

impl_request_has_field!(
    AccountAdjustmentPositionOperation,
    WithAccountAdjustmentPositionOperation,
    operation,
    HasPositionInstrument, position_instrument, &Instrument, instrument;
    HasCollateralAsset, collateral_asset, &Asset, collateral_asset;
);
impl_request_has_field!(
    AccountAdjustmentPositionOperation,
    WithAccountAdjustmentPositionOperation,
    operation,
    HasAverageEntryPrice, average_entry_price, Price, average_entry_price;
    HasPositionMode, position_mode, PositionMode, mode;
    HasAccountAdjustmentPositionLeverage, position_leverage, Option<Leverage>, leverage;
);
impl_request_has_field_passthrough!(
    WithAccountAdjustmentPositionOperation,
    inner,
    HasBalanceAsset, balance_asset, &Asset;
    HasAccountAdjustmentBalanceAverageEntryPrice, balance_average_entry_price, Option<Price>;
    HasAccountAdjustmentBalance, balance, Option<AdjustmentAmount>;
    HasAccountAdjustmentHeld, held, Option<AdjustmentAmount>;
    HasAccountAdjustmentIncoming, incoming, Option<AdjustmentAmount>;
    HasAccountAdjustmentBalanceUpperBound, balance_upper, Option<PositionSize>;
    HasAccountAdjustmentBalanceLowerBound, balance_lower, Option<PositionSize>;
    HasAccountAdjustmentHeldUpperBound, held_upper, Option<PositionSize>;
    HasAccountAdjustmentHeldLowerBound, held_lower, Option<PositionSize>;
    HasAccountAdjustmentIncomingUpperBound, incoming_upper, Option<PositionSize>;
    HasAccountAdjustmentIncomingLowerBound, incoming_lower, Option<PositionSize>;
);

/// Optional post-adjustment inclusive limits for account adjustment components.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct AccountAdjustmentBounds {
    /// Allowed post-adjustment inclusive upper bound for balance.
    pub balance_upper: Option<PositionSize>,
    /// Allowed post-adjustment inclusive lower bound for balance.
    pub balance_lower: Option<PositionSize>,
    /// Allowed post-adjustment inclusive upper bound for held.
    pub held_upper: Option<PositionSize>,
    /// Allowed post-adjustment inclusive lower bound for held.
    pub held_lower: Option<PositionSize>,
    /// Allowed post-adjustment inclusive upper bound for incoming.
    pub incoming_upper: Option<PositionSize>,
    /// Allowed post-adjustment inclusive lower bound for incoming.
    pub incoming_lower: Option<PositionSize>,
}

/// Adds post-adjustment inclusive limits.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct WithAccountAdjustmentBounds<T> {
    pub inner: T,
    pub bounds: AccountAdjustmentBounds,
}

impl_request_has_field!(
    AccountAdjustmentBounds,
    WithAccountAdjustmentBounds,
    bounds,
    HasAccountAdjustmentBalanceUpperBound, balance_upper, Option<PositionSize>, balance_upper;
    HasAccountAdjustmentBalanceLowerBound, balance_lower, Option<PositionSize>, balance_lower;
    HasAccountAdjustmentHeldUpperBound, held_upper, Option<PositionSize>, held_upper;
    HasAccountAdjustmentHeldLowerBound, held_lower, Option<PositionSize>, held_lower;
    HasAccountAdjustmentIncomingUpperBound, incoming_upper, Option<PositionSize>, incoming_upper;
    HasAccountAdjustmentIncomingLowerBound, incoming_lower, Option<PositionSize>, incoming_lower;
);
impl_request_has_field_passthrough!(
    WithAccountAdjustmentBounds,
    inner,
    HasBalanceAsset, balance_asset, &Asset;
    HasAccountAdjustmentBalanceAverageEntryPrice, balance_average_entry_price, Option<Price>;
    HasPositionInstrument, position_instrument, &Instrument;
    HasCollateralAsset, collateral_asset, &Asset;
    HasAverageEntryPrice, average_entry_price, Price;
    HasPositionMode, position_mode, PositionMode;
    HasAccountAdjustmentPositionLeverage, position_leverage, Option<Leverage>;
    HasAccountAdjustmentBalance, balance, Option<AdjustmentAmount>;
    HasAccountAdjustmentHeld, held, Option<AdjustmentAmount>;
    HasAccountAdjustmentIncoming, incoming, Option<AdjustmentAmount>;
);

#[cfg(test)]
mod tests {
    use super::{
        AccountAdjustmentAmount, AccountAdjustmentBalanceOperation, AccountAdjustmentBounds,
        AccountAdjustmentPositionOperation, WithAccountAdjustmentAmount,
        WithAccountAdjustmentBalanceOperation, WithAccountAdjustmentBounds,
        WithAccountAdjustmentPositionOperation,
    };
    use crate::param::{AdjustmentAmount, Asset, Leverage, PositionMode, PositionSize, Price};
    use crate::{
        HasAccountAdjustmentBalance, HasAccountAdjustmentBalanceAverageEntryPrice,
        HasAccountAdjustmentBalanceLowerBound, HasAccountAdjustmentBalanceUpperBound,
        HasAccountAdjustmentHeld, HasAccountAdjustmentHeldLowerBound,
        HasAccountAdjustmentHeldUpperBound, HasAccountAdjustmentIncoming,
        HasAccountAdjustmentIncomingLowerBound, HasAccountAdjustmentIncomingUpperBound,
        HasAccountAdjustmentPositionLeverage, HasAverageEntryPrice, HasBalanceAsset,
        HasCollateralAsset, HasPositionInstrument, HasPositionMode, Instrument,
    };

    #[test]
    fn direct_trait_access_for_balance_operation() {
        let asset = Asset::new("USD").expect("must be valid");
        let average = Price::from_str("1.25").expect("must be valid");
        let operation = AccountAdjustmentBalanceOperation {
            asset: asset.clone(),
            average_entry_price: Some(average),
        };

        assert_eq!(operation.balance_asset(), Ok(&asset));
        assert_eq!(operation.balance_average_entry_price(), Ok(Some(average)));
    }

    #[test]
    fn direct_trait_access_for_position_operation() {
        let instrument = Instrument::new(
            Asset::new("AAPL").expect("must be valid"),
            Asset::new("USD").expect("must be valid"),
        );
        let collateral = Asset::new("USD").expect("must be valid");
        let leverage = Leverage::from_u16(25).expect("must be valid");

        let operation = AccountAdjustmentPositionOperation {
            instrument: instrument.clone(),
            collateral_asset: collateral.clone(),
            average_entry_price: Price::from_str("100").expect("must be valid"),
            mode: PositionMode::Hedged,
            leverage: Some(leverage),
        };

        assert_eq!(operation.position_instrument(), Ok(&instrument));
        assert_eq!(operation.collateral_asset(), Ok(&collateral));
        assert_eq!(
            operation.average_entry_price(),
            Ok(Price::from_str("100").expect("must be valid"))
        );
        assert_eq!(operation.position_mode(), Ok(PositionMode::Hedged));
        assert_eq!(operation.position_leverage(), Ok(Some(leverage)));
    }

    #[test]
    fn direct_trait_access_for_position_operation_fractional_leverage() {
        let leverage = Leverage::from_f64(100.5).expect("must be valid");
        let operation = AccountAdjustmentPositionOperation {
            instrument: Instrument::new(
                Asset::new("AAPL").expect("must be valid"),
                Asset::new("USD").expect("must be valid"),
            ),
            collateral_asset: Asset::new("EUR").expect("must be valid"),
            average_entry_price: Price::from_str("100").expect("must be valid"),
            mode: PositionMode::Hedged,
            leverage: Some(leverage),
        };

        assert_eq!(operation.position_leverage(), Ok(Some(leverage)));
        assert_eq!(leverage.raw(), 1005);
    }

    #[test]
    fn direct_trait_access_for_amount_and_bounds() {
        let balance =
            AdjustmentAmount::Absolute(PositionSize::from_str("4").expect("must be valid"));
        let held = AdjustmentAmount::Delta(PositionSize::from_str("-1").expect("must be valid"));
        let amount = AccountAdjustmentAmount {
            balance: Some(balance),
            held: Some(held),
            incoming: None,
        };

        assert_eq!(amount.balance(), Ok(Some(balance)));
        assert_eq!(amount.held(), Ok(Some(held)));
        assert_eq!(amount.incoming(), Ok(None));

        let bound = PositionSize::from_str("10").expect("must be valid");
        let bounds = AccountAdjustmentBounds {
            balance_upper: Some(bound),
            balance_lower: None,
            held_upper: None,
            held_lower: None,
            incoming_upper: None,
            incoming_lower: None,
        };

        assert_eq!(bounds.balance_upper(), Ok(Some(bound)));
        assert_eq!(bounds.incoming_lower(), Ok(None));
    }

    #[test]
    fn with_wrappers_preserve_access_chain() {
        let base = WithAccountAdjustmentAmount {
            inner: (),
            amount: AccountAdjustmentAmount {
                balance: Some(AdjustmentAmount::Absolute(
                    PositionSize::from_str("7").expect("must be valid"),
                )),
                held: None,
                incoming: None,
            },
        };

        let with_bounds = WithAccountAdjustmentBounds {
            inner: base,
            bounds: AccountAdjustmentBounds {
                balance_upper: Some(PositionSize::from_str("8").expect("must be valid")),
                balance_lower: None,
                held_upper: None,
                held_lower: None,
                incoming_upper: None,
                incoming_lower: None,
            },
        };

        let with_balance = WithAccountAdjustmentBalanceOperation {
            inner: with_bounds,
            operation: AccountAdjustmentBalanceOperation {
                asset: Asset::new("USD").expect("must be valid"),
                average_entry_price: None,
            },
        };

        assert!(with_balance.balance().expect("must be available").is_some());
        assert!(with_balance
            .balance_upper()
            .expect("must be available")
            .is_some());
        assert_eq!(with_balance.balance_average_entry_price(), Ok(None));

        let wrapped_position = WithAccountAdjustmentPositionOperation {
            inner: with_balance,
            operation: AccountAdjustmentPositionOperation {
                instrument: Instrument::new(
                    Asset::new("AAPL").expect("must be valid"),
                    Asset::new("USD").expect("must be valid"),
                ),
                collateral_asset: Asset::new("USD").expect("must be valid"),
                average_entry_price: Price::from_str("1").expect("must be valid"),
                mode: PositionMode::Netting,
                leverage: None,
            },
        };

        assert_eq!(wrapped_position.position_mode(), Ok(PositionMode::Netting));
        assert_eq!(
            wrapped_position.average_entry_price(),
            Ok(Price::from_str("1").expect("must be valid"))
        );
        assert_eq!(wrapped_position.balance_average_entry_price(), Ok(None));
        assert_eq!(wrapped_position.position_leverage(), Ok(None));
    }

    #[test]
    fn borrowed_values_come_from_original_fields() {
        let instrument = Instrument::new(
            Asset::new("AAPL").expect("must be valid"),
            Asset::new("USD").expect("must be valid"),
        );
        let collateral = Asset::new("EUR").expect("must be valid");
        let position = AccountAdjustmentPositionOperation {
            instrument: instrument.clone(),
            collateral_asset: collateral.clone(),
            average_entry_price: Price::from_str("10").expect("must be valid"),
            mode: PositionMode::Hedged,
            leverage: None,
        };

        assert_eq!(position.position_instrument(), Ok(&instrument));
        assert_eq!(position.collateral_asset(), Ok(&collateral));

        let balance = AccountAdjustmentBalanceOperation {
            asset: collateral.clone(),
            average_entry_price: None,
        };

        assert_eq!(balance.balance_asset(), Ok(&collateral));
    }

    #[test]
    fn outer_amount_wrapper_passthroughs_position_branch_traits() {
        let instrument = Instrument::new(
            Asset::new("AAPL").expect("must be valid"),
            Asset::new("USD").expect("must be valid"),
        );
        let collateral = Asset::new("EUR").expect("must be valid");
        let average = Price::from_str("123").expect("must be valid");
        let leverage = Leverage::from_u16(10).expect("must be valid");
        let balance =
            AdjustmentAmount::Absolute(PositionSize::from_str("2").expect("must be valid"));
        let incoming = AdjustmentAmount::Delta(PositionSize::from_str("1").expect("must be valid"));
        let balance_upper = PositionSize::from_str("5").expect("must be valid");
        let incoming_lower = PositionSize::from_str("-2").expect("must be valid");
        let incoming_upper = PositionSize::from_str("6").expect("must be valid");

        let request = WithAccountAdjustmentAmount {
            inner: WithAccountAdjustmentBounds {
                inner: WithAccountAdjustmentPositionOperation {
                    inner: (),
                    operation: AccountAdjustmentPositionOperation {
                        instrument: instrument.clone(),
                        collateral_asset: collateral.clone(),
                        average_entry_price: average,
                        mode: PositionMode::Hedged,
                        leverage: Some(leverage),
                    },
                },
                bounds: AccountAdjustmentBounds {
                    balance_upper: Some(balance_upper),
                    balance_lower: None,
                    held_upper: None,
                    held_lower: None,
                    incoming_upper: Some(incoming_upper),
                    incoming_lower: Some(incoming_lower),
                },
            },
            amount: AccountAdjustmentAmount {
                balance: Some(balance),
                held: None,
                incoming: Some(incoming),
            },
        };

        assert_eq!(request.balance(), Ok(Some(balance)));
        assert_eq!(request.incoming(), Ok(Some(incoming)));
        assert_eq!(request.position_instrument(), Ok(&instrument));
        assert_eq!(request.collateral_asset(), Ok(&collateral));
        assert_eq!(request.average_entry_price(), Ok(average));
        assert_eq!(request.position_mode(), Ok(PositionMode::Hedged));
        assert_eq!(request.position_leverage(), Ok(Some(leverage)));
        assert_eq!(request.balance_upper(), Ok(Some(balance_upper)));
        assert_eq!(request.incoming_upper(), Ok(Some(incoming_upper)));
        assert_eq!(request.incoming_lower(), Ok(Some(incoming_lower)));
    }

    #[test]
    fn outer_amount_wrapper_passthroughs_balance_branch_traits() {
        let asset = Asset::new("EUR").expect("must be valid");
        let average = Price::from_str("1.12").expect("must be valid");
        let held = AdjustmentAmount::Delta(PositionSize::from_str("-3").expect("must be valid"));
        let incoming =
            AdjustmentAmount::Absolute(PositionSize::from_str("4").expect("must be valid"));
        let balance_lower = PositionSize::from_str("-8").expect("must be valid");
        let held_upper = PositionSize::from_str("9").expect("must be valid");
        let held_lower = PositionSize::from_str("-1").expect("must be valid");

        let request = WithAccountAdjustmentAmount {
            inner: WithAccountAdjustmentBounds {
                inner: WithAccountAdjustmentBalanceOperation {
                    inner: (),
                    operation: AccountAdjustmentBalanceOperation {
                        asset: asset.clone(),
                        average_entry_price: Some(average),
                    },
                },
                bounds: AccountAdjustmentBounds {
                    balance_upper: None,
                    balance_lower: Some(balance_lower),
                    held_upper: Some(held_upper),
                    held_lower: Some(held_lower),
                    incoming_upper: None,
                    incoming_lower: None,
                },
            },
            amount: AccountAdjustmentAmount {
                balance: None,
                held: Some(held),
                incoming: Some(incoming),
            },
        };

        assert_eq!(request.held(), Ok(Some(held)));
        assert_eq!(request.incoming(), Ok(Some(incoming)));
        assert_eq!(request.balance_asset(), Ok(&asset));
        assert_eq!(request.balance_average_entry_price(), Ok(Some(average)));
        assert_eq!(request.balance_lower(), Ok(Some(balance_lower)));
        assert_eq!(request.held_upper(), Ok(Some(held_upper)));
        assert_eq!(request.held_lower(), Ok(Some(held_lower)));
    }
}
